File: TI1/TI1.2 SOAP_DOTNET_SINBDD_GR09/02.CLIESC/CONUNI_SOAP_DOTNET_CLIESC_G09/CONUNI_SOAP_DOTNET_CLIESC_G09/servicio/ClienteConversionSoap.cs
using System;
using System.ServiceModel;
using System.ServiceModel.Channels;
using CONUNI_SOAP_DOTNET_CLIESC_G09.modelo;

namespace CONUNI_SOAP_DOTNET_CLIESC_G09.servicio
{
    /// <summary>
    /// Cliente para consumir el servicio SOAP WCF de conversión
    /// </summary>
    public class ClienteConversionSoap
    {
        private const string URL_SERVICIO = "http://10.92.232.246:8085/Service1.svc";
        private const string NAMESPACE = "http://tempuri.org/";
        private const string INTERFAZ_NOMBRE = "WSConversion";

        private ChannelFactory<IWSConversion> _factory;
        private IWSConversion _proxy;

        /// <summary>
        /// Obtiene un mensaje de error amigable basado en el tipo de excepción
        /// </summary>
        private string ObtenerMensajeErrorAmigable(Exception ex)
        {
            // Detectar errores de conexión
            if (ex is EndpointNotFoundException || 
                ex is CommunicationException ||
                ex.Message.Contains("No había ningún extremo") ||
                ex.Message.Contains("listening") ||
                ex.Message.Contains("no endpoint") ||
                ex.Message.Contains("could not be reached") ||
                ex.Message.Contains("no se pudo conectar"))
            {
                return "El servicio no está respondiendo. Por favor, espera un momento y vuelve a intentar.";
            }

            // Detectar errores de timeout
            if (ex is TimeoutException || ex.Message.Contains("timeout") || ex.Message.Contains("tiempo de espera"))
            {
                return "El servicio tardó demasiado en responder. Por favor, intenta de nuevo.";
            }

            // Otros errores - devolver mensaje genérico más amigable
            return "Ocurrió un error al procesar la solicitud. Por favor, verifica que el servicio esté disponible.";
        }

        public ClienteConversionSoap(string urlServicio = URL_SERVICIO)
        {
            try
            {
                // Crear el binding básico HTTP
                BasicHttpBinding binding = new BasicHttpBinding();
                binding.MaxReceivedMessageSize = 2147483647;
                binding.ReaderQuotas.MaxStringContentLength = 2147483647;

                // Crear el endpoint
                EndpointAddress endpoint = new EndpointAddress(urlServicio);

                // Crear el ChannelFactory
                _factory = new ChannelFactory<IWSConversion>(binding, endpoint);
                _proxy = _factory.CreateChannel();
            }
            catch (Exception ex)
            {
                throw new Exception($"Error al conectar con el servicio SOAP: {ex.Message}", ex);
            }
        }

        // ========== TEMPERATURA ==========

        public ConversionResponse CelsiusAFahrenheit(double celsius)
        {
            try
            {
                double resultado = _proxy.CelsiusAFahrenheit(celsius);
                return new ConversionResponse
                {
                    ValorOriginal = celsius,
                    ValorConvertido = resultado,
                    UnidadOrigen = "celsius",
                    UnidadDestino = "fahrenheit",
                    Categoria = "temperatura",
                    Exito = true,
                    Mensaje = "Conversión exitosa"
                };
            }
            catch (Exception ex)
            {
                return new ConversionResponse
                {
                    Exito = false,
                    Mensaje = ObtenerMensajeErrorAmigable(ex)
                };
            }
        }

        public ConversionResponse FahrenheitACelsius(double fahrenheit)
        {
            try
            {
                double resultado = _proxy.FahrenheitACelsius(fahrenheit);
                return new ConversionResponse
                {
                    ValorOriginal = fahrenheit,
                    ValorConvertido = resultado,
                    UnidadOrigen = "fahrenheit",
                    UnidadDestino = "celsius",
                    Categoria = "temperatura",
                    Exito = true,
                    Mensaje = "Conversión exitosa"
                };
            }
            catch (Exception ex)
            {
                return new ConversionResponse
                {
                    Exito = false,
                    Mensaje = ObtenerMensajeErrorAmigable(ex)
                };
            }
        }

        public ConversionResponse CelsiusAKelvin(double celsius)
        {
            try
            {
                double resultado = _proxy.CelsiusAKelvin(celsius);
                return new ConversionResponse
                {
                    ValorOriginal = celsius,
                    ValorConvertido = resultado,
                    UnidadOrigen = "celsius",
                    UnidadDestino = "kelvin",
                    Categoria = "temperatura",
                    Exito = true,
                    Mensaje = "Conversión exitosa"
                };
            }
            catch (Exception ex)
            {
                return new ConversionResponse
                {
                    Exito = false,
                    Mensaje = ObtenerMensajeErrorAmigable(ex)
                };
            }
        }

        public ConversionResponse KelvinACelsius(double kelvin)
        {
            try
            {
                double resultado = _proxy.KelvinACelsius(kelvin);
                return new ConversionResponse
                {
                    ValorOriginal = kelvin,
                    ValorConvertido = resultado,
                    UnidadOrigen = "kelvin",
                    UnidadDestino = "celsius",
                    Categoria = "temperatura",
                    Exito = true,
                    Mensaje = "Conversión exitosa"
                };
            }
            catch (Exception ex)
            {
                return new ConversionResponse
                {
                    Exito = false,
                    Mensaje = ObtenerMensajeErrorAmigable(ex)
                };
            }
        }

        public ConversionResponse FahrenheitAKelvin(double fahrenheit)
        {
            try
            {
                double resultado = _proxy.FahrenheitAKelvin(fahrenheit);
                return new ConversionResponse
                {
                    ValorOriginal = fahrenheit,
                    ValorConvertido = resultado,
                    UnidadOrigen = "fahrenheit",
                    UnidadDestino = "kelvin",
                    Categoria = "temperatura",
                    Exito = true,
                    Mensaje = "Conversión exitosa"
                };
            }
            catch (Exception ex)
            {
                return new ConversionResponse
                {
                    Exito = false,
                    Mensaje = ObtenerMensajeErrorAmigable(ex)
                };
            }
        }

        public ConversionResponse KelvinAFahrenheit(double kelvin)
        {
            try
            {
                double resultado = _proxy.KelvinAFahrenheit(kelvin);
                return new ConversionResponse
                {
                    ValorOriginal = kelvin,
                    ValorConvertido = resultado,
                    UnidadOrigen = "kelvin",
                    UnidadDestino = "fahrenheit",
                    Categoria = "temperatura",
                    Exito = true,
                    Mensaje = "Conversión exitosa"
                };
            }
            catch (Exception ex)
            {
                return new ConversionResponse
                {
                    Exito = false,
                    Mensaje = ObtenerMensajeErrorAmigable(ex)
                };
            }
        }

        // ========== LONGITUD ==========

        public ConversionResponse MetrosAPies(double metros)
        {
            try
            {
                double resultado = _proxy.MetrosAPies(metros);
                return new ConversionResponse
                {
                    ValorOriginal = metros,
                    ValorConvertido = resultado,
                    UnidadOrigen = "metros",
                    UnidadDestino = "pies",
                    Categoria = "longitud",
                    Exito = true,
                    Mensaje = "Conversión exitosa"
                };
            }
            catch (Exception ex)
            {
                return new ConversionResponse
                {
                    Exito = false,
                    Mensaje = ObtenerMensajeErrorAmigable(ex)
                };
            }
        }

        public ConversionResponse PiesAMetros(double pies)
        {
            try
            {
                double resultado = _proxy.PiesAMetros(pies);
                return new ConversionResponse
                {
                    ValorOriginal = pies,
                    ValorConvertido = resultado,
                    UnidadOrigen = "pies",
                    UnidadDestino = "metros",
                    Categoria = "longitud",
                    Exito = true,
                    Mensaje = "Conversión exitosa"
                };
            }
            catch (Exception ex)
            {
                return new ConversionResponse
                {
                    Exito = false,
                    Mensaje = ObtenerMensajeErrorAmigable(ex)
                };
            }
        }

        public ConversionResponse MetrosAPulgadas(double metros)
        {
            try
            {
                double resultado = _proxy.MetrosAPulgadas(metros);
                return new ConversionResponse
                {
                    ValorOriginal = metros,
                    ValorConvertido = resultado,
                    UnidadOrigen = "metros",
                    UnidadDestino = "pulgadas",
                    Categoria = "longitud",
                    Exito = true,
                    Mensaje = "Conversión exitosa"
                };
            }
            catch (Exception ex)
            {
                return new ConversionResponse
                {
                    Exito = false,
                    Mensaje = ObtenerMensajeErrorAmigable(ex)
                };
            }
        }

        public ConversionResponse PulgadasAMetros(double pulgadas)
        {
            try
            {
                double resultado = _proxy.PulgadasAMetros(pulgadas);
                return new ConversionResponse
                {
                    ValorOriginal = pulgadas,
                    ValorConvertido = resultado,
                    UnidadOrigen = "pulgadas",
                    UnidadDestino = "metros",
                    Categoria = "longitud",
                    Exito = true,
                    Mensaje = "Conversión exitosa"
                };
            }
            catch (Exception ex)
            {
                return new ConversionResponse
                {
                    Exito = false,
                    Mensaje = ObtenerMensajeErrorAmigable(ex)
                };
            }
        }

        public ConversionResponse KilometrosAMillas(double kilometros)
        {
            try
            {
                double resultado = _proxy.KilometrosAMillas(kilometros);
                return new ConversionResponse
                {
                    ValorOriginal = kilometros,
                    ValorConvertido = resultado,
                    UnidadOrigen = "kilometros",
                    UnidadDestino = "millas",
                    Categoria = "longitud",
                    Exito = true,
                    Mensaje = "Conversión exitosa"
                };
            }
            catch (Exception ex)
            {
                return new ConversionResponse
                {
                    Exito = false,
                    Mensaje = ObtenerMensajeErrorAmigable(ex)
                };
            }
        }

        public ConversionResponse MillasAKilometros(double millas)
        {
            try
            {
                double resultado = _proxy.MillasAKilometros(millas);
                return new ConversionResponse
                {
                    ValorOriginal = millas,
                    ValorConvertido = resultado,
                    UnidadOrigen = "millas",
                    UnidadDestino = "kilometros",
                    Categoria = "longitud",
                    Exito = true,
                    Mensaje = "Conversión exitosa"
                };
            }
            catch (Exception ex)
            {
                return new ConversionResponse
                {
                    Exito = false,
                    Mensaje = ObtenerMensajeErrorAmigable(ex)
                };
            }
        }

        // ========== PESO ==========

        public ConversionResponse KilogramosALibras(double kilogramos)
        {
            try
            {
                double resultado = _proxy.KilogramosALibras(kilogramos);
                return new ConversionResponse
                {
                    ValorOriginal = kilogramos,
                    ValorConvertido = resultado,
                    UnidadOrigen = "kilogramos",
                    UnidadDestino = "libras",
                    Categoria = "peso",
                    Exito = true,
                    Mensaje = "Conversión exitosa"
                };
            }
            catch (Exception ex)
            {
                return new ConversionResponse
                {
                    Exito = false,
                    Mensaje = ObtenerMensajeErrorAmigable(ex)
                };
            }
        }

        public ConversionResponse LibrasAKilogramos(double libras)
        {
            try
            {
                double resultado = _proxy.LibrasAKilogramos(libras);
                return new ConversionResponse
                {
                    ValorOriginal = libras,
                    ValorConvertido = resultado,
                    UnidadOrigen = "libras",
                    UnidadDestino = "kilogramos",
                    Categoria = "peso",
                    Exito = true,
                    Mensaje = "Conversión exitosa"
                };
            }
            catch (Exception ex)
            {
                return new ConversionResponse
                {
                    Exito = false,
                    Mensaje = ObtenerMensajeErrorAmigable(ex)
                };
            }
        }

        public ConversionResponse GramosAOnzas(double gramos)
        {
            try
            {
                double resultado = _proxy.GramosAOnzas(gramos);
                return new ConversionResponse
                {
                    ValorOriginal = gramos,
                    ValorConvertido = resultado,
                    UnidadOrigen = "gramos",
                    UnidadDestino = "onzas",
                    Categoria = "peso",
                    Exito = true,
                    Mensaje = "Conversión exitosa"
                };
            }
            catch (Exception ex)
            {
                return new ConversionResponse
                {
                    Exito = false,
                    Mensaje = ObtenerMensajeErrorAmigable(ex)
                };
            }
        }

        public ConversionResponse OnzasAGramos(double onzas)
        {
            try
            {
                double resultado = _proxy.OnzasAGramos(onzas);
                return new ConversionResponse
                {
                    ValorOriginal = onzas,
                    ValorConvertido = resultado,
                    UnidadOrigen = "onzas",
                    UnidadDestino = "gramos",
                    Categoria = "peso",
                    Exito = true,
                    Mensaje = "Conversión exitosa"
                };
            }
            catch (Exception ex)
            {
                return new ConversionResponse
                {
                    Exito = false,
                    Mensaje = ObtenerMensajeErrorAmigable(ex)
                };
            }
        }


        {
            try
            {
                return new ConversionResponse
                {
                    ValorConvertido = resultado,
                    Exito = true,
                    Mensaje = "Conversión exitosa"
                };
            }
            catch (Exception ex)
            {
                return new ConversionResponse
                {
                    Exito = false,
                    Mensaje = ObtenerMensajeErrorAmigable(ex)
                };
            }
        }

        {
            try
            {
                return new ConversionResponse
                {
                    ValorConvertido = resultado,
                    Exito = true,
                    Mensaje = "Conversión exitosa"
                };
            }
            catch (Exception ex)
            {
                return new ConversionResponse
                {
                    Exito = false,
                    Mensaje = ObtenerMensajeErrorAmigable(ex)
                };
            }
        }

        {
            try
            {
                return new ConversionResponse
                {
                    ValorConvertido = resultado,
                    UnidadDestino = "onzasFluidas",
                    Exito = true,
                    Mensaje = "Conversión exitosa"
                };
            }
            catch (Exception ex)
            {
                return new ConversionResponse
                {
                    Exito = false,
                    Mensaje = ObtenerMensajeErrorAmigable(ex)
                };
            }
        }

        {
            try
            {
                return new ConversionResponse
                {
                    ValorOriginal = onzasFluidas,
                    ValorConvertido = resultado,
                    UnidadOrigen = "onzasFluidas",
                    Exito = true,
                    Mensaje = "Conversión exitosa"
                };
            }
            catch (Exception ex)
            {
                return new ConversionResponse
                {
                    Exito = false,
                    Mensaje = ObtenerMensajeErrorAmigable(ex)
                };
            }
        }

        // ========== ÁREA ==========

        {
            try
            {
                return new ConversionResponse
                {
                    ValorOriginal = metrosCuadrados,
                    ValorConvertido = resultado,
                    UnidadOrigen = "metrosCuadrados",
                    UnidadDestino = "piesCuadrados",
                    Exito = true,
                    Mensaje = "Conversión exitosa"
                };
            }
            catch (Exception ex)
            {
                return new ConversionResponse
                {
                    Exito = false,
                    Mensaje = ObtenerMensajeErrorAmigable(ex)
                };
            }
        }

        {
            try
            {
                return new ConversionResponse
                {
                    ValorOriginal = piesCuadrados,
                    ValorConvertido = resultado,
                    UnidadOrigen = "piesCuadrados",
                    UnidadDestino = "metrosCuadrados",
                    Exito = true,
                    Mensaje = "Conversión exitosa"
                };
            }
            catch (Exception ex)
            {
                return new ConversionResponse
                {
                    Exito = false,
                    Mensaje = ObtenerMensajeErrorAmigable(ex)
                };
            }
        }

        {
            try
            {
                return new ConversionResponse
                {
                    ValorConvertido = resultado,
                    Exito = true,
                    Mensaje = "Conversión exitosa"
                };
            }
            catch (Exception ex)
            {
                return new ConversionResponse
                {
                    Exito = false,
                    Mensaje = ObtenerMensajeErrorAmigable(ex)
                };
            }
        }

        {
            try
            {
                return new ConversionResponse
                {
                    ValorConvertido = resultado,
                    Exito = true,
                    Mensaje = "Conversión exitosa"
                };
            }
            catch (Exception ex)
            {
                return new ConversionResponse
                {
                    Exito = false,
                    Mensaje = ObtenerMensajeErrorAmigable(ex)
                };
            }
        }

        public void Dispose()
        {
            try
            {
                if (_factory != null)
                {
                    if (_proxy != null && _proxy is ICommunicationObject)
                    {
                        try
                        {
                            ((ICommunicationObject)_proxy).Close();
                        }
                        catch
                        {
                            ((ICommunicationObject)_proxy).Abort();
                        }
                    }
                    _factory.Close();
                }
            }
            catch { }
        }
    }
}

File: Proyectos/TI.1 PAO202650_SINBDD_NRC30742/TI.1.2. PAO202650_SOAP_DOTNET_SINBDD_NRC30742/04. CLIMOV/CONUNI_SOAP_DOTNET_CLIMOV_G05/app/src/main/java/ec/edu/monster/servidor/ClienteConversionSoap.kt
package ec.edu.monster.servidor

import android.util.Log
import ec.edu.monster.modelo.Conversion
import org.ksoap2.SoapEnvelope
import org.ksoap2.serialization.SoapObject
import org.ksoap2.serialization.SoapPrimitive
import org.ksoap2.serialization.SoapSerializationEnvelope
import org.ksoap2.transport.HttpTransportSE

/**
 * CLIENTE SOAP (Modelo de Servicio)
 *
 * Este 'object' (Singleton) es el responsable de conectarse
 * al servidor WCF (.NET) y obtener los datos.
 *
 * Devuelve objetos del tipo 'Conversion' (tu modelo de datos).
 */
object ClienteConversionSoap {

    // --- Configuración Esencial ---
    private const val TAG = "ClienteSOAP"
    private const val NAMESPACE = "http://tempuri.org/"

    private const val SERVER_IP = "10.40.25.94"
    private const val SERVER_PORT = "8085"
    
    // URL del servicio SOAP para emulador Android local.
    //private const val SERVER_IP = "192.168.100.2"
    //private const val SERVER_PORT = "62533"
    private const val URL = "http://$SERVER_IP:$SERVER_PORT/Service1.svc"
    
    private const val INTERFAZ_NOMBRE = "WSConversion" // (De [ServiceContract(Name = ...)])

    // -----------------------------------------------------------------
    // --- FUNCIÓN GENÉRICA PRIVADA (El "Motor") ---
    // -----------------------------------------------------------------

    /**
     * Motor genérico para llamar a cualquier método de conversión del servicio SOAP.
     * Maneja la creación del sobre, la llamada y el parseo de la respuesta.
     *
     * @param methodName El nombre exacto del [OperationContract] (ej. "celsiusAFahrenheit").
     * @param paramName El nombre exacto del parámetro en C# (ej. "celsius").
     * @param paramValue El valor (Double) a enviar.
     * @return Un objeto 'Conversion' con todos los datos del resultado.
     */
    private suspend fun callConversionMethod(methodName: String, paramName: String, paramValue: Double): Conversion {

        // 1. Crear el objeto 'Conversion' para almacenar todo el proceso
        val conversion = Conversion(
            valorOriginal = paramValue,
            operacion = methodName,
            unidadOriginal = paramName, // Asignamos la unidad original
            unidadConvertida = determinarUnidadConvertida(methodName) // Asignamos la unidad de destino
        )

        // --- Configuración de KSOAP2 ---
        val soapAction = "$NAMESPACE$INTERFAZ_NOMBRE/$methodName"

        val request = SoapObject(NAMESPACE, methodName).apply {
            // Convertimos a String para evitar el error "Cannot serialize"
            addProperty(paramName, paramValue.toString())
        }

        val envelope = SoapSerializationEnvelope(SoapEnvelope.VER11).apply {
            dotNet = true // ¡Crítico para WCF .NET!
            setOutputSoapObject(request)
        }

        val transport = HttpTransportSE(URL, 60000) // Timeout de 60 segundos

        // 2. Ejecutar la llamada
        try {
            Log.d(TAG, "Llamando a $methodName con valor $paramValue...")
            transport.call(soapAction, envelope)

            // 3. Procesar la Respuesta (Éxito)
            val response = envelope.response as SoapPrimitive
            val result = response.toString().toDouble()

            // Llenar el objeto 'Conversion' con el éxito
            conversion.valorConvertido = result
            conversion.exitosa = true
            Log.d(TAG, "Respuesta de $methodName: $result")

        } catch (e: Exception) {
            Log.e(TAG, "Error en la llamada a $methodName: ${e.message}", e)

            // 4. Procesar la Respuesta (Error)
            // Llenar el objeto 'Conversion' con el error
            conversion.exitosa = false
            conversion.mensajeError = e.message ?: "Error desconocido"
        }

        // 5. Devolver el objeto 'Conversion' completo
        return conversion
    }

    /**
     * Función helper para asignar la unidad de destino basado en el nombre del método
     */
    private fun determinarUnidadConvertida(methodName: String): String {
        return when {
            // Temperatura
            methodName.contains("Fahrenheit") -> "fahrenheit"
            methodName.contains("Celsius") -> "celsius"
            methodName.contains("Kelvin") -> "kelvin"
            // Longitud
            methodName.contains("Pies") -> "pies"
            methodName.contains("Metros") -> "metros"
            methodName.contains("Pulgadas") -> "pulgadas"
            methodName.contains("Millas") -> "millas"
            methodName.contains("Kilometros") -> "kilometros"
            // Peso
            methodName.contains("Libras") -> "libras"
            methodName.contains("Kilogramos") -> "kilogramos"
            methodName.contains("Onzas") -> "onzas"
            methodName.contains("Gramos") -> "gramos"


            else -> "desconocida"
        }
    }

    // -----------------------------------------------------------------
    // --- MÉTODOS PÚBLICOS (La API que usará tu App) ---
    // --- Ahora todos devuelven el objeto 'Conversion' ---
    // -----------------------------------------------------------------

    // --- Temperatura ---
    suspend fun celsiusAFahrenheit(celsius: Double) =
        callConversionMethod("celsiusAFahrenheit", "celsius", celsius)

    suspend fun fahrenheitACelsius(fahrenheit: Double) =
        callConversionMethod("fahrenheitACelsius", "fahrenheit", fahrenheit)

    suspend fun celsiusAKelvin(celsius: Double) =
        callConversionMethod("celsiusAKelvin", "celsius", celsius)

    suspend fun kelvinACelsius(kelvin: Double) =
        callConversionMethod("kelvinACelsius", "kelvin", kelvin)

    suspend fun fahrenheitAKelvin(fahrenheit: Double) =
        callConversionMethod("fahrenheitAKelvin", "fahrenheit", fahrenheit)

    suspend fun kelvinAFahrenheit(kelvin: Double) =
        callConversionMethod("kelvinAFahrenheit", "kelvin", kelvin)

    // --- Longitud ---
    suspend fun metrosAPies(metros: Double) =
        callConversionMethod("metrosAPies", "metros", metros)

    suspend fun piesAMetros(pies: Double) =
        callConversionMethod("piesAMetros", "pies", pies)

    suspend fun metrosAPulgadas(metros: Double) =
        callConversionMethod("metrosAPulgadas", "metros", metros)

    suspend fun pulgadasAMetros(pulgadas: Double) =
        callConversionMethod("pulgadasAMetros", "pulgadas", pulgadas)

    suspend fun kilometrosAMillas(kilometros: Double) =
        callConversionMethod("kilometrosAMillas", "kilometros", kilometros)

    suspend fun millasAKilometros(millas: Double) =
        callConversionMethod("millasAKilometros", "millas", millas)

    // --- Peso/Masa ---
    suspend fun kilogramosALibras(kilogramos: Double) =
        callConversionMethod("kilogramosALibras", "kilogramos", kilogramos)

    suspend fun librasAKilogramos(libras: Double) =
        callConversionMethod("librasAKilogramos", "libras", libras)

    suspend fun gramosAOnzas(gramos: Double) =
        callConversionMethod("gramosAOnzas", "gramos", gramos)

    suspend fun onzasAGramos(onzas: Double) =
        callConversionMethod("onzasAGramos", "onzas", onzas)

}
